### PR TITLE
go/roothash: Change runtime round timeout to be in blocks

### LINKED ISSUE #3210: Runtime round timeout should be in blocks instead of time
In order to simplify the timeout processing logic and avoid spurious timeouts when consensus stalls, runtime round timeouts should be in blocks instead of time.

### PR DESCRIPTION
**Pull request by kostko**
Fixes #3210

### CODE REVIEW

File: go/oasis-node/cmd/registry/runtime/runtime.go [hunk 1]
```diff
@@ -537,7 +537,7 @@ func init() {
 	runtimeFlags.Uint64(CfgExecutorGroupSize, 1, "Number of workers in the runtime executor group/committee")
 	runtimeFlags.Uint64(CfgExecutorGroupBackupSize, 0, "Number of backup workers in the runtime executor group/committee")
 	runtimeFlags.Uint64(CfgExecutorAllowedStragglers, 0, "Number of stragglers allowed per round in the runtime executor group")
-	runtimeFlags.Duration(CfgExecutorRoundTimeout, 10*time.Second, "Executor committee round timeout for this runtime")
+	runtimeFlags.Int64(CfgExecutorRoundTimeout, 5, "Executor committee round timeout for this runtime (in consensus blocks)")
```
Comment (ptrus) [inline go/oasis-node/cmd/registry/runtime/runtime.go:540]: Maybe add an extra note to the `.changelog/3210.breaking.md`:

> Note: Existing deployments will need to alter the state dump and update the executor round timeout in all registered compute runtimes.

File: go/roothash/api/commitment/pool.go [hunk 2]
```diff
@@ -420,9 +421,9 @@ func (p *Pool) TryFinalize(
 	defer func() {
 		if rearmTimer {
 			// All timeouts are rounded to nearest second to ensure stable serialization.
```
Comment (ptrus) [inline go/roothash/api/commitment/pool.go:423]: Outdated comment

File: go/consensus/tendermint/apps/roothash/roothash.go [hunk 3]
```diff
@@ -580,13 +526,49 @@ func (app *rootHashApplication) tryFinalizeBlock(
...
+		}
+	}(rtState.ExecutorPool.NextTimeout)
+
 	finalizedBlock := app.tryFinalizeExecutor(ctx, rtState, forced)
 	if finalizedBlock == nil {
 		return nil
 	}
 
-	if err := app.postProcessFinalizedBlock(ctx, rtState, finalizedBlock); err != nil {
+	if err = app.postProcessFinalizedBlock(ctx, rtState, finalizedBlock); err != nil {
```
Comment (ptrus) [inline go/consensus/tendermint/apps/roothash/roothash.go:571]: hm i don't think the `app.postProcessFinalizedBlock` can even ever return an error, maybe update the function definition.

Also the initial error check can then be skipped in the deffered function

File: go/oasis-node/cmd/registry/runtime/runtime.go (hunk 1 above)
Comment (kostko) [reply go/oasis-node/cmd/registry/runtime/runtime.go:540]: Good idea, added.

**Comment by ptrus** [comment]
I think it should be also pretty simple adding a "TestRoundTimeout" test to the `roothash/test/tester` for this (we currently only test successful round there):

Something like:
- get latest roothash block
- wait for runtime.Exectuor.RoundTimeout consensus blocks
- ensure that timeout happened and round failed

(just added a similar test in my "proposer timeouts", that's why i know :-) )

File: go/oasis-node/cmd/registry/runtime/runtime.go [hunk 4]
```diff
@@ -372,7 +372,7 @@ func runtimeFromFlags() (*registry.Runtime, signature.Signer, error) { // nolint
 			GroupSize:         viper.GetUint64(CfgExecutorGroupSize),
 			GroupBackupSize:   viper.GetUint64(CfgExecutorGroupBackupSize),
 			AllowedStragglers: viper.GetUint64(CfgExecutorAllowedStragglers),
-			RoundTimeout:      viper.GetDuration(CfgExecutorRoundTimeout),
+			RoundTimeout:      viper.GetInt64(CfgExecutorRoundTimeout),
```
Comment (ptrus) [inline go/oasis-node/cmd/registry/runtime/runtime.go:375]: not related to this PR but i also just noticed that we don't verify any (other than `GroupSize`) of the `ExecutorParameters` when registering a runtime.

Comment (ptrus) [reply go/oasis-node/cmd/registry/runtime/runtime.go:375]: Specifically for `RoundTimeout` should probably allow only `0<=X<=max`, `max` mostly so we don't need to worry about overflowing. Maybe also not allow non sensible small values such us `<5`?

Comment (kostko) [reply go/oasis-node/cmd/registry/runtime/runtime.go:375]: Validation for RoundTimeout (actually exactly for "<5" :grin:) has already been added in this PR?

Comment (ptrus) [reply go/oasis-node/cmd/registry/runtime/runtime.go:375]: Oh, sorry missed that :-) maybe also impose a sensible max limit? (and add validation for other executor params that miss it / or we can open an issue for the rest)

Comment (kostko) [reply go/oasis-node/cmd/registry/runtime/runtime.go:375]: Refactored the runtime descriptor validity checks a bit, please take another look.

Comment (ptrus) [reply go/oasis-node/cmd/registry/runtime/runtime.go:375]: looks good :+1:

**Review by ptrus — APPROVED** [review]
Looks good (there's one more minor pending comment)

File: go/roothash/api/commitment/pool.go [hunk 5]
```diff
@@ -445,6 +445,11 @@ func (p *Pool) TryFinalize(
 			// process the round, assuming that it is possible to do so.
 			if isTimeoutAuthoritative {
 				p.Discrepancy = true
+				// Arm the timer, but increase the roundTimeout as the backup workers should be
+				// given some more time to do the computation.
+				rearmTimer = true
+				// roundTimeout = 1.5*roundTimeout
+				roundTimeout = (15 * roundTimeout) / 10
```
Comment (ptrus) [inline go/roothash/api/commitment/pool.go:452]: maybe move the `15` into a constant

Comment (kostko) [reply go/roothash/api/commitment/pool.go]: Good idea, added numerator/denominator constants.

File: go/consensus/tendermint/apps/roothash/roothash.go (hunk 3 above)
Comment (kostko) [reply go/consensus/tendermint/apps/roothash/roothash.go]: Changed.

**Pull request merged**